sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java

```diff
@@ -0,0 +1,384 @@
+        }
+        String systemVersionWithoutV = removeVFromSystemVersion(compasSystemVersion.get());
+        Optional<TCompasICDHeader> compasICDHeader = PrivateUtils.extractCompasPrivate(tied, TCompasICDHeader.class);
+        if (compasICDHeader.isEmpty()) {
+            return new CriteriaOrError(null, "No private COMPAS-ICDHeader found in this IED");
+        }
+        if (compasICDHeader.get().getIEDSystemVersioninstance() == null) {
+            return new CriteriaOrError(null, "No IEDSystemVersioninstance in the COMPAS-ICDHeader of this IED");
+        }
+        TBayIntOrExt bayIntOrExt = cbName.endsWith("I") ? TBayIntOrExt.BAY_INTERNAL : TBayIntOrExt.BAY_EXTERNAL;
+
+        return new CriteriaOrError(
+                new Criteria(cbType,
+                        systemVersionWithoutV,
+                        TIEDType.fromValue(compasICDHeader.get().getIEDType().value()),
+                        TIEDRedundancy.fromValue(compasICDHeader.get().getIEDredundancy().value()),
+                        compasICDHeader.get().getIEDSystemVersioninstance(),
+                        bayIntOrExt), null);
+    }
+
+    private String removeVFromSystemVersion(TCompasSystemVersion compasSystemVersion) {
+        String[] minorVersionParts = compasSystemVersion.getMinorSystemVersion().split("\\.");
+        return (minorVersionParts.length == 3) ?
+                compasSystemVersion.getMainSystemVersion() + "." + minorVersionParts[0] + "." + minorVersionParts[1]
+                : null;
+    }
+
+    private Criteria vlanToCriteria(TVlan vlan) {
+        requireNotNull(vlan.getCBType(), "CBType");
+        requireNotBlank(vlan.getXY(), "XY");
+        requireNotBlank(vlan.getZW(), "ZW");
+        requireNotNull(vlan.getIEDType(), "IEDType");
+        requireNotNull(vlan.getIEDRedundancy(), "IEDRedundancy");
+        requireNotBlank(vlan.getIEDSystemVersionInstance(), "IEDSystemVersionInstance");
+        requireNotNull(vlan.getBayIntOrExt(), "BayIntOrExt");
+
+        return new Criteria(
+                vlan.getCBType(),
+                vlan.getXY() + "." + vlan.getZW(),
+                vlan.getIEDType(),
+                vlan.getIEDRedundancy(),
+                toIedSystemVersionInstance(vlan.getIEDSystemVersionInstance()),
+                vlan.getBayIntOrExt()
+        );
+    }
+
+    private Settings vlanToSetting(TVlan vlan) {
+        return new Settings(toVLanId(vlan.getVlanId()), toVlanPriority(vlan.getVlanPriority()), toDurationInMilliSec(vlan.getMinTime()), toDurationInMilliSec(vlan.getMaxTime()));
+    }
+
+    private void requireNotBlank(String str, String attribute) {
+        if (StringUtils.isBlank(str)) {
+            throw new ScdException("Error in Control Block communication setting file: vlan is missing attribute " + attribute);
+        }
+    }
+
+    private void requireNotNull(Object o, String attribute) {
+        if (Objects.isNull(o)) {
```

Comment:
I'm unclear about how this method behaves.
We can use Reflection API to verify if an attribute exist or not.

```suggestion
private void requireField(TVlan vlan, String attribute) {
    try {
         vlan.getClass().getDeclaredField(attribute);
     } catch(NoSuchFieldException e) {
          throw new ScdException("Error in Control Block communication setting file: vlan is missing attribute " +      attribute);
      }
```

Reply:
It is like `Objects.requireNonNull(T obj, String message)` but it throws a `ScdException` instead of a `NullPointerException`, and the message is pre-formatted so you juste need to give the attribute name.
I don't want to use reflection here, because I think reflection should be used in last resort.

To make it clearer, I can change the second parameter to `String message` like this :
```java
    private void requireNotNull(Object o, String message) {
        if (Objects.isNull(o)) {
            throw new ScdException(message);
        }
    }
```

and use it like this : 

```java
    private static final String MISSING_ATTRIBUTE = "Error in Control Block communication setting file: vlan is missing attribute ";
...
   requireNotNull(vlan.getCBType(), MISSING_ATTRIBUTE + "CBType");
```

Reply:
It looks clear after changing the second parameter to String message.

Reply:
OK